// @alicloud/fun/lib/language-service/model/arrayASTNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const astNode_1 = require("./astNode");
const jsonParser_1 = require("../parser/jsonParser");
class ArrayASTNode extends astNode_1.ASTNode {
    constructor(parent, start, end) {
        super(parent, 'array', start, end);
        this.items = [];
    }
    getChildNodes() {
        return this.items;
    }
    getValue() {
        return this.items.map(item => item.getValue());
    }
    addItem(item) {
        if (item) {
            this.items.push(item);
            return true;
        }
        return false;
    }
    valdiate(schema, validationResult, matchingSchemas) {
        super.validate(schema, validationResult, matchingSchemas);
        this.items.forEach(item => {
            const itemValidationResult = new jsonParser_1.ValidationResult();
            item.validate(schema.items, itemValidationResult, matchingSchemas);
            validationResult.mergePropertyMatch(itemValidationResult);
        });
    }
}
exports.ArrayASTNode = ArrayASTNode;
